src/libs/ascent/runtimes/ascent_main_runtime.cpp

```diff
@@ -616,24 +615,77 @@ AscentRuntime::EnsureDomainIds()
     }
 
     int domain_offset = 0;
+    int max_num_domains = num_domains;
+
 #ifdef ASCENT_MPI_ENABLED
     int *domains_per_rank = new int[comm_size];
     MPI_Allgather(&num_domains, 1, MPI_INT, domains_per_rank, 1, MPI_INT, mpi_comm);
     for(int i = 0; i < m_rank; ++i)
     {
       domain_offset += domains_per_rank[i];
     }
+    for(int i = 0; i < comm_size; i++)
+    {
+      if(domains_per_rank[i] > max_num_domains)
+        max_num_domains = domains_per_rank[i];
+    }
     delete[] domains_per_rank;
 #endif
+
+    std::unordered_set<int> local_unique_ids;
+
+    int *domain_ids = new int[max_num_domains];
     for(int i = 0; i < num_domains; ++i)
     {
       conduit::Node &dom = m_source.child(i);
 
       if(!dom.has_path("state/domain_id"))
       {
          dom["state/domain_id"] = domain_offset + i;
+         domain_ids[i] = domain_offset + i;
+         local_unique_ids.insert(domain_ids[i]);
       }
+      else
+      {
+         domain_ids[i] = dom["state/domain_id"].to_int32();
+         local_unique_ids.insert(domain_ids[i]);
+      }
+    }
+
+    if(local_unique_ids.size() != num_domains)
+    {
+      ASCENT_ERROR("Local Domain IDs are not unique ");
     }
+
+#ifdef ASCENT_MPI_ENABLED
+
+    int total_domains = 0;
+    MPI_Allreduce(&num_domains, &total_domains, 1, MPI_INT, MPI_SUM, mpi_comm);
+    for(int i = num_domains; i < max_num_domains; i++)
+    {
+      domain_ids[i] = 0;
+    }
+
+    int *domain_ids_per_rank = new int[comm_size*max_num_domains];
```

Comment:
Is max_num_domains - the max domains per mpi rank?

Could we just directly count number of domain trees per rank and alloc exactly? 
We could then:
- init an array to all -1s
- use per mpi task offset and fill with local domain ids
- use an MPI MAX all reduce to get the full set of domain ids

Then we could unordered set (or even an in place sort + scan) to detect dups.

If we are in bad shape, we should provide which rank(s) have duplicate ids in the error message.

Reply:
max_num_domains is the max number of domains of any rank, since there might be a mismatch of number of domains between ranks, I wanted to standardize a passing size. 

But your way works better, I'll change it.